text/0000-relax-mixin-contract.md

```diff
@@ -0,0 +1,69 @@
+
+This RFC proposed that we update the buildpack specification to create a new relationship between `mypkg`, `run:mypkg`, and `build:mypkg`.
+
+- `run:mypkg` and `mypkg` imply the same set of changes to the run image, such that buildpacks requiring `mypkg` would be satisfied with `run:mypkg` at runtime (and vice versa).
+
+- `build:mypkg` and `mypkg` imply the same set of changes to the build image, such that buildpacks requiring `mypkg` would be satisfied with `build:mypkg` at build-time (and vice versa).
+
+- A runtime stackpack that needs to satisfy `run:mypkg` or `mypkg` would only be provided with `mypkg`, regardless of which version is required. The stackpack would have no knowledge of whether it could expect the build image to have the corresponding mixin.
+
+- A build-time stackpack that needs to satisfy `build:mypkg` or `mypkg` would only be provided with `mypkg`, regardless of which version is required. The stackpack would have no knowledge of whether it could expect the run image to have the corresponding mixin.
+
+
+# Definitions
+[definitions]: #definitions
+
+- [**mixins** and **stage specifiers**](https://github.com/buildpacks/spec/blob/main/platform.md#mixins)
+- [**contractual compatibility** driven by ABI](https://github.com/buildpacks/spec/blob/main/platform.md#compatibility-guarantees)
+
+
+# Motivation
+[motivation]: #motivation
+
+1. Stack authors must currently provide both stage-specific and non-stage-specific versions of mixin names on each image to ensure that buildpacks can require mixins with or without the stage specifier. This is unfriendly UX for stack authors, and difficult to explain.
+
+2. Stack authors are unable to relax the contract between the build and run image (by only using the stage-specifier version of the mixins) without requiring the buildpacks manually list the mixin twice with the different stage specifiers.
+
+3. Under the pre-RFC mixin definition, stackpacks (specified in [#87](https://github.com/buildpacks/rfcs/issues/87)) will attempt to re-install packages when stage-specifier is mismatched between the buildpack and stack images. For instance, a buildpack requesting `mypkg` would result in the stackpack attempting to install `mypkg` on a run image with `run:mypkg`.
+
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Pack and lifecycle will relax contractual restrictions when initiating builds, creating builders, and exporting images.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This change reduces the power of the mixin model for stack authors. Previously, stack authors could reliably use `run:mypkg` to indicate that a run image strictly contains the runtime version of mypkg (which may provide different functionality than `mypkg`). Now `run:mypkg` and `mypkg` imply the exact same thing for the run image.
```

Comment:
Would you mind clarifying this point? Are `run:mypkg` and `mypkg` different packages? If yes, why do they satisfy each other, and if not, why would they provide different functionality? A concrete example would be fantastic!

Reply:
I believe they could be different packages. The apt contract we have applies only to the `io.buildpacks.stacks.bionic` stack as per this [RFC](https://github.com/buildpacks/rfcs/blob/main/text/0031-bionic-mixins.md).

Reply:
@sclevine mind chiming in here?

Reply:
The set of changes to the run image represented by `run:mypkg` and `mypkg` are the same, so if `mypkg` represents an OS package in the context of the run image, `run:mypkg` must represent that package as well.